domain-knowledge/BUSINESS-RULES.md

```diff
@@ -0,0 +1,548 @@
+any of the `Artist's` other `Albums`.
+ 
+This logic applies only a per-`Artist` basis; that is, if an `Artist` uploads
+a file that is identical to a file that a _different_ `Artist` has uploaded
+previously, then the identical copy _is_ stored. This rule ensures that `Artists`
+have exclusive ownership over the audio files associated with their `Songs`, but
+that needless duplication is avoided wherever practical. If it were
+possible to share a `FlacFile` between two different `Artists`, seemingly simple
+determinations, such as "Who uploaded this file?", become far more difficult,
+with no real benefit beyond marginal storage gains.
+
+## Account Closure & Deletion
+
+`Account` closure is implemented with "soft-deletion" on the `User` object, so it
+is completely reversible. When a `User` closes an `Account`, all publicly
+accessible content associated with that user disappears from the service
+immediately. 
+
+To reopen the account, the `User` may simply log in again and confirm the intention
+to reopen the account.
+
+Account deletion, on the other hand, is implemented with "hard-deletion", so it
+is permanent and cannot be undone. 
+
+### For Non-Artist/Label
+
+When a `User` that is not associated with an `Artist` or `Label`, the following
+objects are deleted *inside a transaction* (and in this order, to prevent
+foreign key failures):
+
+- `OrderItem`
+- `Order`
+- `Account`
+- `User`
+
+### For Artist/Label
+
+#### For Artist
+
+When a `User` is associated with an `Artist`, additional provisions apply when
+deleting an `Account`.
+
+When an `Artist` requests that its `Account` be deleted, the deletion is queued,
+pending an eligibility check is against the following rules:
+
+1. All albums have been disabled (this prevents customers from making further
+   purchases, which would extend the deletion timeline for 90 more days).
+2. All monies owed to the `Artist` have been paid to its associated `CatalogEntity`.
+3. 90 days have elapsed since a customer last purchased the `Artist's` content.
+4. All of the `Artist's` albums have been deleted.
+5. All of the `Artist's` uploaded images have been deleted.
+
+Once every 24 hours, eligibility is evaluated, and if any of those statements is
+false, the `Account` is ineligible for deletion, and the service will complete
+any steps that it can, automatically, before the next evaluation.
+
+Eventually, all statements will evaluate as true, at which time the `Account`,
+and all associated content (database records and files) is hard-deleted.
+
+Upon successful deletion, the following objects are deleted, *inside a transaction*
+(and in this order, to prevent foreign key failures):
+
+- `OrderItem`
+- `Order`
+- `Profile`
+- `Account`
+- `User`
+
+*Before the transaction is committed*, a confirmation email message is queued to
+be sent to the email address associated with the `User` record. Upon successful
+queueing, the transaction is committed.
```

Comment:
If it is before it is committed wouldn't that mean you could send out an email to say it was deleted but then the transaction fails?

Reply:
Yes, that's absolutely true. If memory serves (and often it does not!), my thinking here is that the user should have confirmation that the deletion was at least attempted. But, to your point, there is probably a more comprehensive way to structure this logic, such that the user is apprised, accurately, of initiation _and_ success (or failure). Let me think on this a bit more (or feel free to offer suggestions).

Reply:
After digging into this a bit more, I now recall my concern:

If all of the database records associated with the user are hard-deleted before the confirmation email is queued, and the queued job fails in a way that re-queuing it is futile, we've essentially "lost" the `User` and have no easy way to notify them  when the failure is resolved and their information is actually deleted.

But, now that I "type this out loud", the process, as currently documented, does not solve this problem. We would have to hold-off on deleting the database records until the queued job is processed _successfully_. And what if that never happens (for whatever unforeseeable reason, e.g., queue is lost/corrupted)? In such a failure scenario, we may not even realize that the `User` tried to delete their `Account`.

It seems that we would need to record the deletion attempt, prior to initiating the transaction, and mark the attempt as completed if the transaction is successful.

But then where in the process does the email notification happen?

The element of complexity here is that `Account` deletion must be _queued_ once requested due to all the prerequisites that must be met, some of which may take time (several days, weeks, or even months).

I suppose that we could send _two_ emails, the first of which is sent immediately upon _requesting_ deletion, and the second of which is sent upon _actual_ deletion.

Or, we could just display a message upon requesting deletion, e.g., "Your request to delete your account has been queued for processing. Once the above prerequisites are met and your account is deleted, a confirmation will be sent to [insert email here]." And then we send only one email, upon actual deletion.

Thoughts?

Reply:
I just pushed a commit that I feel addresses this subject as thoroughly as is reasonably possible. If you have any additional (or especially better) ideas, let's open a separate Issue for this.